Add smoke test for JWT realm wiring

justincr-elastic opened the pull request:
Verify JWT realm settings can be loaded by ES server, and queried by ES client.

Verify REST client can submit JWT and client secret, and ES server JWT realm will perform successful authentication.

justincr-elastic committed:
Initial commit.

justincr-elastic committed:
Graceful handle errors when printing realm settings in tests. If read local file throws an exception, catch and show the message in the print out.

For client secret, use random letters to avoid implying Base64 is required.

For OIDC-compatible HMAC key, use random letters to avoid implying Base64URL is required. Using Base64URL of original bytes is OK too, so randomly choose between them.

justincr-elastic committed:
RealmInfoIT is passing. JwtRealmAuthIT is failed for unknown reason.

Refactored JWT utils to specify static JWT claims instead of random.

Added test to generate JWT realm settings, JWT and client secret, and dump the signed JWT's header and claims set.

Added smoke tests. RealmInfoIT is working. JwtRealmAuthIT should be working based on the same realm config and request headers working in JwtRealmTests, but not working yet for some reason.

justincr-elastic commented:
Hi @ywangd, I think coding is done, but I am running into two issues.

1. JWT string value is much longer than 140 characters. Automatic formatting does not fix it.
`./gradlew       spotlessApply -Des.format.everything checkstyle`

2. I generated JWT realm settings, JWT, and client secret in JwtRealmTests.testCreateJwtSmokeTestRealm(). That test verifies authentication succeeded too. However, when I put those JWT realm settings, JWT, and client secret, in the smoke test it fails to authenticate. I am not getting any server logs, only client logs, so I am unclear why it fails or how to diagnose the issue.

3. I left some debug logs in the code with TODO to remove them. I will after I figure out what is broken in 2.

justincr-elastic commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:
```diff
@@ -0,0 +1,59 @@
+    private static final String ROLE_NAME = "security_test_role";
+    private static final String HEADER_CLIENT_SECRET = "client-shared-secret-string";
+    private static final String HEADER_JWT =
+        "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE";
```
Spotless doesn't know how to fix this. I tried putting it in a file and loading it, but I get read access denied.

Is there another way to keep it intact, without breaking it up just to satisfy checkstyle?

justincr-elastic commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:
```diff
@@ -0,0 +1,59 @@
+                + HEADER_CLIENT_SECRET
+        );
+        LOGGER.info(JwtRealm.HEADER_END_USER_AUTHENTICATION + ": " + JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME + " " + HEADER_JWT);
+        LOGGER.info(options.build().toString());
```
I expected the test to succeeded. I will remove these logs after we figure out why the test is failing.

justincr-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:
```diff
@@ -107,16 +107,18 @@ public JwtRealm(final RealmConfig realmConfig, final SSLService sslService, fina
-        this.jwksAlgsPkc = this.parseJwksAlgsPkc(false); // reloadable
+        this.parseJwksAlgsHmac();
+        this.parseJwksAlgsPkc();
+        this.verifyAnyAvailableJwkAndAlgPair();
```
Logging bug fix. If HMAC key filtered out (ex: too small), diagnostic log in PKC JWKSet method was being skipped because it was empty. I moved the checks to a separate method to ensure it gets printed. I found this issue while trying to generate the static realm config for the smoke test.

justincr-elastic commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java:
```diff
@@ -108,6 +104,122 @@ public void shutdown() throws Exception {
+     * If something needs tweaking, change the code here and re-run it. The new artifacts can be used to update that smoke test project.
+     * @throws Exception Error if something goes wrong.
+     */
+    public void testCreateJwtSmokeTestRealm() throws Exception {
```
The output of this test is the static JWT realm config, client secret, and signed JWT, all to be used in the realm config smoke test. The output of this test is saved in the README.asciidoc. The result of this test is deterministic, re-running it produces the same realm config, client secret, and signed JWT.

justincr-elastic commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java:
```diff
@@ -328,15 +378,21 @@ public static OctetSequenceKey randomJwkHmacOidc(final JWSAlgorithm jwsAlgorithm
-        return utf8HmacKeyBuilder.build();
+        final String passwordKey;
+        if (randomBoolean()) {
+            final Base64URL hmacKeyBytesBase64 = hmacKey.getKeyValue(); // Random bytes => 8 bits/byte search space
```
I previously used Base64 of random bytes as a new plain text string. It worked, but gave misleading impressing base64 was required. It is not, so I added some fuzzing to ensure random letter passwords can be used too.

justincr-elastic commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java:
```diff
@@ -363,48 +419,28 @@ public static OctetSequenceKey jwkHmacRemoveAttributes(final OctetSequenceKey hm
-    }
-
-    public static SignedJWT randomValidJwsHeaderAndJwtClaimsSet(
+    public static SignedJWT buildUnsignedJwt(
```
Refactored to pass in all values, instead of generating some as random. This was required to generate a reproducible JWT for the realm config smoke test, with fixed claims like exp claim time 2099/1/1 12:00:00.000.

justincr-elastic committed:
Revert JWT realm files. Moved them to separate PR.

justincr-elastic committed:
Update docs/changelog/84249.yaml

elasticsearchmachine commented:
Hi @justincr-elastic, I've created a changelog YAML for you.

ywangd commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:
```diff
@@ -0,0 +1,59 @@
+    private static final String ROLE_NAME = "security_test_role";
+    private static final String HEADER_CLIENT_SECRET = "client-shared-secret-string";
+    private static final String HEADER_JWT =
+        "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE";
```
We can break this long string into a concatenation, e.g. 
```
        final String HEADER_JWT = "eyJhbGciOiJIUzI1NiJ9."
            + "eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZX"
            + "MiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA"
            + "5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE";
```

justincr-elastic replied:
Resolved. I did something similar.

ywangd commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/README.asciidoc:
```diff
@@ -0,0 +1,57 @@
+  realms=[1], expected=[jwt8,8], current[jwt8,8], status=[SUCCESS], authenticated=[true], msg=[null], meta=[{}], user=[User[username=security_test_user,roles=[security_test_role],fullName=null,email=null,metadata={jwt_claim_iss=iss8, jwt_claim_sub=security_test_user, jwt_claim_aud=[aud8], jwt_claim_roles=[security_test_role]}]].
+[2022-02-24T01:17:47,884][INFO ][o.e.x.s.a.j.JwtRealmTests] [testCreateJwtSmokeTestRealm] Test succeeded
+[2022-02-24T01:17:47,887][INFO ][o.e.x.s.a.j.JwtRealmTests] [testCreateJwtSmokeTestRealm]
+===
+Request Headers
+===
+X-Client-Authentication: client-shared-secret-string
+Authorization: eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE
+
+===
+elasticsearch.yml settings
+===
+xpack.security.authc.realms.jwt.jwt8.order: 8
+xpack.security.authc.realms.jwt.jwt8.allowed_issuer: iss8
+xpack.security.authc.realms.jwt.jwt8.allowed_signature_algorithms: [HS256]
+xpack.security.authc.realms.jwt.jwt8.allowed_audiences: [aud8]
+xpack.security.authc.realms.jwt.jwt8.claims.principal: sub
+xpack.security.authc.realms.jwt.jwt8.client_authentication.type: SharedSecret
+
+===
+PKC JWKSet contents
+===
+Not found.
+
+===
+elasticsearch-keystore secure settings
+===
+xpack.security.authc.realms.jwt.jwt8.hmac_key: hmac-oidc-key-string-for-hs256-algorithm
+xpack.security.authc.realms.jwt.jwt8.client_authentication.shared_secret: client-shared-secret-string
+
```
I think we only need this section in this file. Rest is noise and it's easier to understand the point without it.

justincr-elastic replied:
Resolved.

ywangd commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/clientsecret.txt:
```diff
@@ -0,0 +1 @@
+client-shared-secret-string
```
I don't think this file is needed. right?

justincr-elastic replied:
Resolved. Removed file.

ywangd commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/jwt.txt:
```diff
@@ -0,0 +1 @@
+eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE
```
Same here. This file is not needed?

justincr-elastic replied:
Resolved. Removed file.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java:
```diff
@@ -717,7 +717,7 @@ private void multipleRealmsAuthenticateJwtHelper(
                 assertThat(new TreeSet<>(Arrays.asList(user.roles())), equalTo(new TreeSet<>(Arrays.asList(authenticatedUser.roles()))));
                 if (jwtIssuerAndRealm.realm.delegatedAuthorizationSupport.hasDelegation()) {
                     assertThat(user.metadata(), is(equalTo(authenticatedUser.metadata()))); // delegated authz returns user's
-                                                                                            // metadata
+                    // metadata
```
Err can we revert this change. The format looks off to me. Or make the trailing comment into its own line.

justincr-elastic replied:
Resolved. I have no idea what that comment wrapped to the next line. Fixed it.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:
```diff
@@ -409,6 +409,11 @@
         "oidc",
         License.OperationMode.PLATINUM
     );
+    public static final LicensedFeature.Persistent JWT_REALM_FEATURE = LicensedFeature.persistent(
+        REALMS_FEATURE_FAMILY,
+        "jwt",
+        License.OperationMode.PLATINUM
+    );
```
I guess you haven't update the PR yet because this file needs the extra Rest Header change.

justincr-elastic replied:
Resolved. Yes, when we found the issue it close to midnight for me. I added the request header to the accept list first thing next morning in commit 0a50dc9e5d534ed1a4a02cce1d9e68e80472a20c.

ywangd reviewed (commented):
This PR looks promising. I have a few comments besides what we discussed in synchronously. Please request review again once you make it formal. Thanks!

justincr-elastic committed:
Remove client test logs.

justincr-elastic committed:
Add JWT client authc header to accept list.

justincr-elastic committed:
Merge remote-tracking branch 'fork/feature/jwt-realm-tech-preview' into feature/jwt-realm-tech-preview

justincr-elastic committed:
Only check authc in JwtRealmAuthIT, like PkiRealmAuthIT. Change authz check to empty roles.

It is possible to add role mapping with a template to the "roles" claim in the JWT. However, if PKI doesn't need to check it, then JWT doesn't need to check it either.

justincr-elastic committed:
Revert.

justincr-elastic commented on .idea/checkstyle-idea.xml:
```diff
@@ -16,4 +16,4 @@
     </option>
   </component>
-</project>
+</project>
```
Resolved.

justincr-elastic committed:
Cleanup README.asciidoc

justincr-elastic committed:
Remove two unnecessary text files.

justincr-elastic committed:
Fix comment so it doesn't wrap to next line

justincr-elastic committed:
Add JWT to a second realm config smoke test.

justincr-elastic commented:
I am curious why are there two smoke tests for realm config?
- x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/RealmInfoIT.java
- x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/SecurityRealmSettingsTests.java

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

tvernum commented:
I think we can consider `SecurityRealmSettingsTests` to be obsolete. The QA test is better because it actually starts a real node.
I think it's worth opening an issue to review whether the QA test covers everything that is in `SecurityRealmSettingsTests` and then delete `SecurityRealmSettingsTests`.

ywangd commented on x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:
```diff
@@ -0,0 +1,49 @@
+        final Map<String, Object> authenticate = super.authenticate(options);
+        assertUsername(authenticate, USERNAME);
+        assertRealm(authenticate, "jwt", "jwt8");
+        assertRoles(authenticate); // empty list
```
So we decided to go without role mapping? I am ok with it since this smoke test is more about the realm itself. Just to be sure.

justincr-elastic replied:
Correct, I went without role mapping. This PR was just about wiring in the config.

I considered spending more cycles during this PR to add a one-off authz check. However, Tim had recommended I create smaller, incremental PRs to keep the scope of each one tight. So, this PR is just for checking the config is wired in. The next planned PR is for integration testing, and that feels like a better place to verify role mapping for JWT realm.

ywangd reviewed (approved):
LGTM

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-tech-preview

justincr-elastic commented:
@elasticmachine update branch

Jenkins shutdown killed my test. Need to rerun it again.

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-tech-preview

justincr-elastic commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/jwt-realm-tech-preview

justincr-elastic commented:
@elasticmachine run elasticsearch-ci/part-1
